### DISCUSSION TITLE
Plugin config spread and multiple config states

### DISCUSSION THREAD

**Discussion by cbratschi** [payloadcms/payload#10008] [Feature Requests & Ideas]
Payload's plugin architecture has a design flaw. As outlined by the official specs, spread operators shall be used to modify the config in a plugin:

https://payloadcms.com/docs/plugins/build-your-own#spread-syntax

This leads to immutable config versions which are passed from plugin to plugin to create another one. Plugins often use closures and store a certain version known at the runtime of the plugin but this version differs from the final config after all plugins got executed.

For instance I had the issue today that the nested-docs plugin failed detecting drafts here in a collection hook:

https://github.com/payloadcms/payload/blob/main/packages/plugin-nested-docs/src/hooks/resaveSelfAfterCreate.ts#L23

```TypeScript
const updateAsDraft =
        typeof collection.versions === 'object' &&
        collection.versions.drafts &&
        doc._status !== 'published'
```

In my case I moved adding collection.versions.drafts to a new plugin which was executed after the nested-docs plugin and nested-docs used an older config version here:

https://github.com/payloadcms/payload/blob/main/packages/plugin-nested-docs/src/index.ts#L57

It's difficult to difficult to fix this issue without changing the architecture from ground up. At least this issue should be part of the specification and plugins should avoid passing older configs to their hooks. These cases are very difficult to debug, it just fails. Re-ordering plugins might help but first the root cause has to be known.

I created this discussion to find a solution for upcoming Payload versions.

**Comment by AlessioGr** [comment]
You're 100% right here. We'll go ahead and fix the usage of config data in the nested docs plugin hooks.

I don't think there is a way to _guarantee_ that this does not happen anymore and it's up to the plugin to do it right. To address this, we'll implement the following changes in our documentation

- Discourage unnecessary use of spread operators: For performance reasons and to prevent creating multiple copies of the config
- Caution against passing config data from the plugin in closures: Instead, it should be accessed through `req.payload.config`

**Comment by cbratschi** [comment]
I rechecked this against current Payload 4 `main`. The original nested-docs failure is now fixed: PR [#13629](https://github.com/payloadcms/payload/pull/13629) changed the hooks to use their runtime `collection` argument instead of a collection captured during plugin execution.

The underlying contract is unchanged, though. `buildConfig` still passes the result of each plugin to the next plugin, so the `config` received by a plugin is an intermediate value. The new `definePlugin` API helps with ordering and cross-plugin options, but its `config` argument is not the final runtime config either.

The documentation warning proposed above has not yet been added. Both the basic and advanced plugin-authoring pages should explain that hooks and other delayed callbacks must not close over config, collection, or global objects from plugin execution. They should resolve final state through runtime hook arguments or `req.payload.config`.

I can prepare a documentation PR and a focused nested-docs regression test covering a later plugin enabling drafts. Would that be useful?
